common/src/main/java/io/druid/concurrent/LifecycleLock.java

```diff
@@ -0,0 +1,238 @@
+ * under the License.
+ */
+
+package io.druid.concurrent;
+
+import java.util.concurrent.locks.AbstractQueuedSynchronizer;
+
+/**
+ * A synchronization tool for lifecycled objects (see {@link io.druid.java.util.common.lifecycle.Lifecycle}, that need
+ * happens-before between start() and other methods and/or to check that the object was successfully started in other
+ * methods.
+ *
+ * Guarantees in terms of JMM: happens-before between {@link #exitStart()} and {@link #isStarted()},
+ * exitStart() and {@link #canStop()}, if it returns {@code true}.
+ *
+ * Example:
+ * class ExampleLifecycledClass
+ * {
+ *   final LifecycleLock lifecycleLock = new LifecycleLock();
+ *
+ *   void start()
+ *   {
+ *     if (!lifecycleLock.canStart()) {
+ *       .. return or throw exception
+ *     }
+ *     try {
+ *       .. do start
+ *       lifecycleLock.started();
+ *     }
+ *     finally {
+ *       lifecycleLock.exitStart();
+ *     }
+ *   }
+ *
+ *   void otherMethod()
+ *   {
+ *     Preconditions.checkState(lifecycleLock.isStarted());
+ *     // all actions done in start() are visible here
+ *     .. do stuff
+ *   }
+ *
+ *   void stop()
+ *   {
+ *     if (!lifecycleLock.canStop()) {
+ *       .. return or throw exception
+ *     }
+ *     // all actions done in start() are visible here
+ *     .. do stop
+ *   }
+ * }
+ */
+public final class LifecycleLock
+{
+  private static class Sync extends AbstractQueuedSynchronizer
+  {
+    private static final int NOT_STARTED = 0;
+    private static final int STARTING = 1;
+    private static final int STARTED = 2;
+    private static final int START_EXITED_SUCCESSFUL = 3;
+    private static final int START_EXITED_FAIL = 4;
+    private static final int STOPPING = 5;
+    private static final int STOPPED = 6;
+
+    boolean canStart()
+    {
+      return compareAndSetState(NOT_STARTED, STARTING);
+    }
+
+    void started()
+    {
+      if (!compareAndSetState(STARTING, STARTED)) {
+        throw new IllegalMonitorStateException("Called started() not in the context of start()");
+      }
+    }
+
+    void exitStart()
+    {
+      // see tryReleaseShared()
+      releaseShared(1);
+    }
+
+    @Override
+    protected boolean tryReleaseShared(int ignore)
+    {
+      while (true) {
+        int state = getState();
+        if (state == STARTING) {
+          if (compareAndSetState(STARTING, START_EXITED_FAIL)) {
+            return true;
+          }
+        } else if (state == STARTED) {
+          if (compareAndSetState(STARTED, START_EXITED_SUCCESSFUL)) {
+            return true;
+          }
+        } else {
+          throw new IllegalMonitorStateException("exitStart() called not in the end of the start() method");
+        }
+      }
+    }
+
+    boolean isStarted()
+    {
+      try {
+        // see tryAcquireShared()
+        acquireSharedInterruptibly(1);
+      }
+      catch (InterruptedException e) {
+        throw new RuntimeException(e);
```

Comment:
I didn't see we do with InterruptedExceptions anything but propagation and logging, in the end of the day. I feel making `awaitStarted()` to throw `InterruptedException` will just add a lot of boilerplate and no benefit.

There is also a non-interruptible method in `AbstractQueuedSynchronizer`, `acquireShared()`, I didn't use it because decided that it's always better to be able to be interrupted, to avoid deadlock in some cases.